runtime/src/accounts_db.rs

```diff
@@ -2209,67 +2251,186 @@ impl AccountsDb {
+                        num_acceptable_failed_iterations += 1;
+                    }
+                }
+                LoadedAccountAccessor::Cached(None) => {
+                    // Cache was flushed in between checking the index and retrieving from the cache,
+                    // so retry.
+                    num_acceptable_failed_iterations += 1;
+                    if is_root_fixed {
+                        // it's impossible for this to fail for transaction loads more than once.
+                        // This is because:
+                        // 1) For a slot `X` that's being replayed, there is only one
+                        // latest ancestor containing the latest update for the account, and this
+                        // ancestor can only be flushed once.
+                        // 2) The root cannot move while replaying, so the index cannot continually
+                        // find more up to date entries than the current `slot`
+                        assert!(num_acceptable_failed_iterations <= 1);
+                    }
+                }
+                LoadedAccountAccessor::Cached(Some(_)) | LoadedAccountAccessor::Stored(Some(_)) => {
+                    // Everything else means there was no race, so break out and continue
+                    break;
+                }
+            }
+            let (new_slot, new_store_id, new_offset) =
+                self.accounts_index_get(ancestors, pubkey, max_root)?;
+            if new_slot == slot && new_store_id == store_id {
+                // If the entry was missing from the cache, that means it must have been flushed,
+                // and the accounts index is always updated before cache flush, so this
+                // should be impossible
+                assert!(new_store_id != CACHE_VIRTUAL_STORAGE_ID);
+
+                // If this is not a cache entry, then this was a minor fork slot
+                // that had its storage entries cleaned up by purge_slots() but hasn't been
+                // cleaned yet, so return None
+                return None;
+            }
+
+            slot = new_slot;
+            store_id = new_store_id;
+            offset = new_offset;
+
+            // Accounts cache was just flushed, look in the index to find
+            // the account storage again. This works because in accounts cache flush,
+            // an account is written to storage *before* it is removed from the cache
+            account_accessor =
+                self.get_account_accessor_from_cache_or_storage(slot, pubkey, store_id, offset);
+        }
+
+        Some((account_accessor, slot))
     }
 
-    pub fn load_slow(&self, ancestors: &Ancestors, pubkey: &Pubkey) -> Option<(Account, Slot)> {
-        self.load(ancestors, pubkey)
+    fn do_load(
+        &self,
+        ancestors: &Ancestors,
+        pubkey: &Pubkey,
+        max_root: Option<Slot>,
+        // true if the root will not move during this load
```

Comment:
maybe better of defining this `move`? like `(i.e. not flushed from AccountsCache nor cleaned and shrinked)` (Yeah, I know this is a bit verbose... But I want some documented definition of this flag in accessible way for future me).

Reply:
> Also, this is really only `root`? I think random AccountsCache evinction and full-of-cache condition could flush the un rooted ones too? so, this could more be named like `is_slot_fixed`?

This is true, but based on the *usage* of that flag, I couldn't come up with a better name than `is_root_fixed`, see comment here: https://github.com/solana-labs/solana/pull/15454/files#r590921257